app/src/main/kotlin/io/intrepid/skotlinton/screens/example2/Example2ViewModel.kt

```diff
@@ -9,29 +9,29 @@ import io.reactivex.rxkotlin.subscribeBy
 
 class Example2ViewModel(configuration: ViewModelConfiguration) : BaseViewModel(configuration) {
 
-    val currentIpAddressText = BehaviorRelay.create<String>()
-    val previousIpAddressVisible = BehaviorRelay.create<Boolean>()
-    val previousIpAddressText = BehaviorRelay.create<String>()
+    val currentIpAddressText = MutableLiveData<String>()
```

Comment:
I originally considered separating out the public immutable version from the private mutable version, ex
```
private val _currentIpAddressText = MutableLiveData<String>()
val currentIpAddressText: LiveData<String> get() = _currentIpAddressText
```
but decided against it since it adds too much boilerplate just to prevent the view from accidentally calling setValue/postValue. It would be nice if something like https://github.com/Kotlin/KEEP/pull/122 is implemented.

Reply:
This is a case where practicality should be valued over the theory. It's just too tedious to introduce 2 variables for every livedata (the IDE helps a bit with the Introduce backing property intention, but there are still extra work). It also clutters up the viewModel, leads to weird code style where some of the private properties have underscore prefix while some don't, and could actually be harder to maintain since AS doesn't help with refactoring.

It's not worth all these hassle just to prevent a developer from accidentally modifying the livedata outside of the viewmodel.

Reply:
@f2rf2r 
You could add this to the `BaseViewModel` implementation:
```
    protected var <T : Any> LiveData<T>.latestValue: T?
        get() = (this as MutableLiveData<T>).value
        set(value) {
            (this as MutableLiveData<T>).value = value
        }
```
Then you can declare exposed live-data properties as plain `LiveData<?>`, as long as you assign them a value of `MutableLiveData<?>` and only a ViewModel (that is subclassed from `BaseViewModel`) can change its value by setting its extended property `latestValue`.

Reply:
this is a very good solution to the issue. thanks!

Reply:
It would have been best if `LiveData` was an interface to begin with and not exposing the `value` either (an Activity/Fragment should never directly access the `value` of a `LiveData`, only `observe` it). But this is how LiveData was designed by Google.... you can create your own interface that looks very similar to LiveData, but does not expose `value` at all.

Reply:
Having language level support for this (https://github.com/Kotlin/KEEP/pull/122) would be great too since this pattern would appear in a bunch of other places/classes. ex: List vs MutableList vs ArrayList

Reply:
C++ has this through `friend` relationships. It comes in handy, but confuses things as well.

If only Kotlin had 'package private', where you'd define your ViewModels in one package and your Activities/Fragments in another....